lib/asn1.ml

```diff
@@ -28,17 +28,36 @@ struct
     }
     [@@deriving ord,eq,show]
 
+    let fix_negative_rsa = function
+      | Ok {n; e} ->
+        if Z.gt n Z.zero then
+          Ok { n; e }
+        else
+          (* A few tools generate keys without leading 0 byte, which gets interpreted
+             as a negative integer. This fix should account for these bad keys. *)
+          let length = Z.numbits n + 1 in
+          let all_fs = String.make (length / 4) 'f' |> Z.of_string_base 16 in
+          let positive_n = Z.add all_fs n in
+          if Z.gt positive_n Z.zero then
+            Ok { n = positive_n; e}
+          else
```

Comment:
Can this ever happen?

Reply:
This could happen if we misjudge the length of the binary field required for the operation. I don't have a firm mathematical proof one way or the other.